Add feature to set test iteration count from command line

bjackman opened the pull request:
This is a kludge, but I think doing this properly basically requires
implementing our own Nose test runner, or even our own test-running
framework entire. In the short term we really need this feature.

nosetests doesn't provide a way for the user to pass command-line
arguments to the test. To get around this, this commit adds a
parameter to lisa-test that sets an environment variable, which
LisaTest then reads to set or override the iterations count.

---

This will go pretty nicely with https://github.com/ARM-software/lisa/pull/348 but they are technically independent.

bjackman commented:
> We already changed the source so this is a no-brainer.

Huh? What source?

derkling commented on src/shell/lisa_shell:
```diff
@@ -308,19 +319,38 @@ nosetests -v --nocapture --nologcapture \
+if [[ $have_more_args == no ]]; then
+        # No arguments provided, default to "help"
+        _lisa-test-usage
+else
```
return -1 and remove this else?

derkling commented on src/shell/lisa_shell:
```diff
@@ -308,19 +319,38 @@ nosetests -v --nocapture --nologcapture \
+                        shift
+                        ;;
+                *)
+                        have_more_args=yes
```
What's that used for?
Can't we go for the usual
```bash
while getopts ...; do
  ...
done
```
and move the following if into the default case?

bjackman replied:
Yeah I did have a rationale for doing it this way but I can't remember what it was so it was probably nonsense :D

bjackman replied:
Although I'd like to keep it as "raw bash" because `getopts` doesn't support longopts like "--iterations". [There's a separate thing called `getopt` but it isn't in OSX etc](https://stackoverflow.com/questions/402377/using-getopts-in-bash-shell-script-to-get-long-and-short-command-line-options) and I guess one day someone might want to get LISA working on Macs. And the raw bash is fine in this case I think.

derkling commented on src/shell/lisa_shell:
```diff
@@ -277,9 +277,12 @@ fi
 function _lisa-test-usage {
 cat <<EOF
-Usage: lisa-test [args] FILE[:CLASS]
+Usage: lisa-test [--wload-iterations ITERATIONS] [args] FILE[:CLASS]
```
why not saving @sinkap some typing and name this parameter just `iterations`?

bjackman replied:
The reason I went for `--wload-iterations` was because the thing that's repeated is each workload and not the test as a whole... but actually I guess that's pretty much the same thing so will change it.

derkling replied:
But does it really make sense to run a test multiple times?
Eventually, and that's what we do, we run a workload multiple time and aggregate results to define if THE test pass or not... thus there should not be ambiguity and I'm pretty sure we will unlikely need in the future to run A test multiple times.

derkling replied:
Is this going to be changed or you still prefer to keep this name?

bjackman replied:
Agh, sorry I thought I changed it.

derkling commented on libs/utils/test.py:
```diff
@@ -87,6 +87,14 @@ def runExperiments(cls):
 
         experiments_conf = cls._getExperimentsConf(test_env)
+
+        if ITERATIONS_FROM_CMDLINE:
```
Why not naming this just ITERATIONS which matches the name used for the experiments_conf attribute?

bjackman replied:
Because I want it to be as clear as possible that it's a different source for the value.. the mechanism I've implemented here pretty convoluted so I think long descriptive names are gonna be helpful..

derkling replied:
Well, we already have a big warning on console to notify the user about the override.

Moreover, I would not be so negative about considering this an hack... nosetest does not support command line parameters and thus, either we can add support for them or use environment variable. An environment variable is still a legal way to defined parameters for an app... ;-)

Moreover, to be honest, I don't think we will even update this code in the near future... mainly because that's doing exactly what we need and it perfectly works for what we need.

However, if you are more comfortable with this name... let's go with it! ;-)

derkling commented on libs/utils/test.py:
```diff
@@ -227,4 +235,20 @@ def experiment_test(wrapped_test, instance, args, kwargs):
+# immediately instead of going ahead with target setup etc.
+try:
+    ITERATIONS_FROM_CMDLINE = int(
+        os.getenv('LISA_TEST_ITERATIONS_COUNT', '0'))
```
Here instead it should probable be just `LISA_TEST_ITERATIONS`

bjackman commented:
- Changed the param to "--iterations"
- "LISA_TEST_ITERATIONS_COUNT" -> "LISA_TEST_ITERATIONS"
- Simplified the arg-parsing loop as you suggested

but

- Didn't use `getopts`
- Didn't rename "ITERATIONS_FROM_CMDLINE"

derkling commented on src/shell/lisa_shell:
```diff
@@ -308,19 +319,35 @@ nosetests -v --nocapture --nologcapture \
-esac
-echo
+local iterations=0 # Means use default - see libs/utils/test.py
+local have_more_args=no
```
Is this still required?

bjackman replied:
Nope, removing, thanks

derkling commented on src/shell/lisa_shell:
```diff
@@ -308,19 +319,35 @@ nosetests -v --nocapture --nologcapture \
+                        iterations="$2"
+                        shift
+                        ;;
+                *)
```
Perhaps I'm note really getting what you wanna do here.
Isn't this extremely fragile?

AFAIU we are forced to put `-n/--interations `before the remaining params, otherwise we will pass it to the `__list_test` instead and run just 1 iteration.

Did not tried but I still have the impression that getopts can be useful to make this more robust... maybe I'll give a try later. Can you just post here an example of how you wanna call a test with parameters and multiple iterations?

bjackman replied:
Hmm yeah that's true this doesn't work if you want to pass extra args to nosetests - that's probably what my rationale for the previous implementation was. Will see what I can come up with...

bjackman commented:
> But does it really make sense to run a test multiple times?
Eventually, and that's what we do, we run a workload multiple time and aggregate results to define if THE test pass or not... thus there should not be ambiguity and I'm pretty sure we will unlikely need in the future to run A test multiple times.

[Can't reply directly to comment]

Yep, I agree, I meant to change "--wload-iterations" to "--iterations" but missed out the docstring in the commit, will fix.

bjackman commented:
OK pushed a new version with hopefully a more robust parsing strategy that just picks up "-n" or "--iterations" and passes everything else through to nosetests. Can't figure out how to do this with getopts..

bjackman commented:
OK, updated again fixing the doc for "--iterations" parameter.. hopefully I didn't miss anything this time.

ionela-voinescu commented:
Except for some nitpicks it looks good to me. I'll start some tests of both PRs (https://github.com/ARM-software/lisa/pull/415 and https://github.com/ARM-software/lisa/pull/348) overnight and make changes to CI after these are merged. Good job and thank you for the changes. They are very useful!
